### PR TITLE
Handle batches with only a single empty sentence

### PR DESCRIPTION
**Pull request by schmmd**
Fixes https://github.com/allenai/allennlp/issues/837

This also fixes some type annotations.

### CODE REVIEW

**Comment by DeNeutoy** [comment]
My 2 cents: If you accidentally added a new line where there wasn't one, you want to 1) be warned about it and 2) have this elmo dump not crash and 3) not have a tensor filled with zeros to represent a `\n` character.

Instead, it should warn the user that they provided an empty line, print the line number and remove it from the embedding.

**Comment by matt-peters** [comment]
What Mark said.  Since we are keying the file on the sentence string, it does no harm to ignore empty sentences, and a vector of all zeros in the output file is strange.

**Comment by schmmd** [comment]
@DeNeutoy my concern is that I want the user to be able to `zip(embeddings, sentences)`.  If I slip a sentence out, then they cannot rely on this working and I need to change the contracts.  Any suggestions here?

I might have created the wrong tensor.  My goal was to create a tensor that represents all the embeddings for the input.  In this case the input had 0 words, so that tensor would have shape `(3, 0, 1024)`.

I'm happy to additionally strip empty lines out of `embed_file`--I think that would be a good change since the output is a dict.

**Comment by schmmd** [comment]
@matt-peters it does no harm in `embed_file` (and I'll make that change) but it does harm in `embed_sentence` (I do not want to make that change there).

**Comment by matt-peters** [comment]
Right, I see you point now.  Maybe returning `None` instead of the `(3, 0, 1024)` tensor is better?

**Comment by schmmd** [comment]
@matt-peters why would returning `None` be better?  I think you'll still get a `(3, 0, 1024)` tensor if you have an empty input as part of a batch--I'm just handling the case where a batch has a single empty sentence (which currently crashes).

**Comment by schmmd** [comment]
I would love to move that special-case handling for batches with a single empty sentence lower in AllenNLP--but I don't know how to.

**Comment by schmmd** [comment]
I pushed a change that strips empty lines from the input file.  @matt-peters if you feel strongly about it, I can change the code to warn when there are empty lines instead of ignoring them.

**Comment by matt-peters** [comment]
Makes sense now, your original approach is good.  We can look at the single empty sentence issue maybe down the road.

File: allennlp/commands/elmo.py [hunk 1]
```diff
@@ -211,15 +214,26 @@ def embed_batch(self, batch: List[List[str]]) -> List[torch.Tensor]:
         """
         elmo_embeddings = []
 
-        embeddings, mask = self.batch_to_embeddings(batch)
-        for i in range(len(batch)):
-            length = int(mask[i, :].sum())
-            sentence_embeds = embeddings[i, :, :length, :].data.cpu().numpy()
-            elmo_embeddings.append(sentence_embeds)
+        # Batches will only an empty sentence will throw an exception inside AllenNLP, but rather than throw an
```
Comment (DeNeutoy) [inline allennlp/commands/elmo.py]: This sentence is a bit messed up.

File: allennlp/commands/elmo.py [hunk 2]
```diff
@@ -211,15 +214,26 @@ def embed_batch(self, batch: List[List[str]]) -> List[torch.Tensor]:
...
+                # Slicing the embedding :0 throws an exception so we need to special case for empty sentences.
+                if length == 0:
+                    elmo_embeddings.append(empty_embedding())
+                else:
+                    elmo_embeddings.append(embeddings[i, :, :length, :].data.cpu().numpy())
 
         return elmo_embeddings
 
-    def embed_sentences(self, sentences: Iterable[List[str]], batch_size: int) -> Iterable[torch.Tensor]:
+    def embed_sentences(
```
Comment (DeNeutoy) [inline allennlp/commands/elmo.py]: We'd normally format this as 
```
def embed_sentences(self,
                    param1: type,
                    param2: type)
```

**Comment by DeNeutoy** [comment]
Actually, I think this line is now incorrect: https://github.com/allenai/allennlp/pull/842/files#diff-cca0c8911d8dcaf0850e4be6d0a669e4R280

`split_sentences` is no longer guaranteed to be the same length as `sentences`, if you have extra new lines, and `zip([1,2,3], [1,2]) == [(1,1), (2,2)]`, so if you have empty lines, the hdf5 keys will be  for the wrong tensors.

**Comment by schmmd** [comment]
@DeNeutoy -- you're right, I made that change in the wrong place.  Fix oncoming.

**Comment by matt-gardner** [comment]
Can you also add a test that makes sure this case is handled correctly?

**Comment by schmmd** [comment]
Actually, scratch my last comment.  I got very confused by the GitHub diff.  I'll look closer.

**Comment by schmmd** [comment]
Yep--that's my starting point ;-)

**Comment by schmmd** [comment]
Mark and I discussed a bit offline, but neither of us moved in our opinions and Mark didn't want to discuss it more.  I'm going to merge this now, but I'm happy to discuss it more as there's a lot I have yet to learn about deep learning, and while my approach might make sense for me, it might not make sense to others.  Just open an issue and assign me.

The present state is that:

* When processing a file, empty lines are stripped.  I'm happy to add a warning for this case and when there are duplicates if anyone would like such a warning--just cut an issue and assign me.

* In `embed_sentence` empty sentences are returned with shape `(3, 0, 1024)` instead of throwing an exception, as they did before.

```
$ ee.embed_sentence("This is".split()).shape
(3, 2, 1024)

$ ee.embed_sentence("This".split()).shape
(3, 1, 1024)

$ ee.embed_sentence("".split()).shape
(3, 0, 1024)
```

**Pull request merged**